homeassistant/components/tradfri/__init__.py

```diff
@@ -143,7 +143,7 @@ async def async_keep_alive(now: datetime) -> None:
         gw_status = True
         try:
             await api(gateway.get_gateway_info())
-        except RequestError:
+        except PytradfriError:
```

Comment:
What stack trace could occur here? The api only raises `RequestError`.

Reply:
The lib is to be considered a black box, so we cannot assume what it raises or not raises !

Already today it raises more than requestError, it also raises the timeout and probably more coming from the underlaying lib. Those are all, right now, inherited from RequestError, but there are no guarantee it continues to be like that. (actually the PR I am preparing break that pattern).

Reply:
The current library design has all api errors inherit from `RequestError`. That's intentional and not a black box.

Until the library changes design I suggest we keep the current scope of exception catching.

Reply:
Actually that is not total correct ColorError is inherited from PytradfriError and can as far as I can see be raised in a faulty request (see group and light_control).

Reply:
That's a good example why we don't want to increase the scope of the exception caught here.

`ColorError` is raised when the library user passes an incorrect color to `set_predefined_color`. We wouldn't want to catch that exception at this place (if we would have used `set_predefined_color`) since it would have masked a bug and incorrect usage of the library. If we would allow the user to pass arbitrary color names to `set_predefined_color` we should catch `ColorError` as close to that part of our code as possible, eg in a service handler.

Reply:
You are still using knowledge of how the library operates internally, and remembering the library is not only for HA (as you told me a couple of times in PRs for the library), we should not use such knowledge here.

I had exactly the same problem in modbus with pymodbus, but strangely enough there you had no problems approving the change.

As a codeowner/maintainer I believe this is the right choise to make, you as a reviewer should control if I violate HA implementation rules. The only looser here are the stability of tradfri, unless you want to maintain tradfri. I try to make a stable product (4 more to cone).

Reply:
The library exceptions are not only for internal use. They are meant to be imported and used by library users. Instead we should document the design of the library exceptions to make them clear.

Reply:
but that is not in the scope here, that should be done in the library, and by using Pytradfrierror we secure the integration.

I can follow your line of thought if the integration were to handle the exceptions differently, but it is not (nor have you suggested that).

I still see the PR as correctly following HA written as well as unwritten rules…And I strongly believe this PR makes tradfri easier to maintain.

Please tell me why you apply the restriction here but accepted the ver same changes in modbus. If it is because you want to decide on the future of tradfri, the simply say it, and I will stop.

Reply:
No, our standard is to use the smallest scope necessary when catching exceptions, only catch expected exceptions and exceptions that make sense in the local context, so that we don't mask bugs.

I don't know what you're referring to in modbus, sorry.

Reply:
Considering your statement “RequestError” is also not to be used, because that is NOT the smallest scope. RequestError is like PytradfriError a sum of exceptions, and therefore by far not the smallest scope !

Furthermore RequestError and PytradfriError are for all purposes in this integration identical, so you cannot claim that PytradfriError is broader scope the RequestError.

Reply:
Using PytradfriError is actually better, because it show the origin directly, whereas with RequestError you need (especially in config_flow.py) to check the import.

Reply:
`RequestError` represents the exceptions that the tradfri API calls may raise. If we want to handle all of those the same way it's ok to catch `RequestError`.

`RequestError` is not the only exception in pytradfri as you pointed out so we should not catch `PytradfriError` here as other exceptions are not expected and catching broader may mask bugs.

I hope I've made myself clear. I don't know how to better explain this.

Reply:
You have for a long time made yourself clear. I do agree with your explanation, but you base these findings on your internal knowledge of how the library works and not documented facts. The approach is dangerous, since others with less knowledge might change either the integration or the library without you reviewing it.

You do not remember modbus, probably because you do not have the same in-depth knowledge of pymodbus, and simply accepted my judgement as codeowner, which just goes to prove my point.

But no need for more discussion, you do not want to see it as two independent repos.

Followup PR will be submitted within the hour.